File: packages/shared-kmp/features/auth/src/commonMain/kotlin/dev/forcetower/melon/feature/auth/domain/repository/PasskeyRepository.kt
package dev.forcetower.melon.feature.auth.domain.repository

import dev.forcetower.melon.core.common.Outcome
import dev.forcetower.melon.feature.auth.domain.model.PasskeyAttestation
import dev.forcetower.melon.feature.auth.domain.model.PasskeyCredential
import dev.forcetower.melon.feature.auth.domain.model.PasskeyError
import dev.forcetower.melon.feature.auth.domain.model.PasskeyRegistrationOptions

// Manages the signed-in account's passkeys against apps/api: lists them,
// fetches WebAuthn creation options, enrolls a freshly minted credential, and
// renames/revokes existing ones. Distinct from `AuthRepository`, which owns
// the pre-session login (assertion) flow.
internal interface PasskeyRepository {
    suspend fun registrationOptions(): Outcome<PasskeyRegistrationOptions, PasskeyError>

    suspend fun register(
        attestation: PasskeyAttestation,
        deviceName: String?,
    ): Outcome<Unit, PasskeyError>

    suspend fun list(): Outcome<List<PasskeyCredential>, PasskeyError>

    suspend fun rename(id: String, deviceName: String): Outcome<Unit, PasskeyError>

    suspend fun delete(id: String): Outcome<Unit, PasskeyError>
}

File: apps/ios/UNESKit/Sources/UNESKit/Domain/Repositories/PasskeyRepository.swift
import ComposableArchitecture

/// Manages the account's passkeys against apps/api: lists them, fetches
/// registration options, and enrolls a freshly minted credential.
@DependencyClient
struct PasskeyRepository: Sendable {
    var credentials: @Sendable () async throws -> [PasskeyCredential]
    var registrationOptions: @Sendable () async throws -> PasskeyRegistrationOptions
    var register: @Sendable (_ attestation: PasskeyAttestation, _ deviceName: String?) async throws -> Void
    var rename: @Sendable (_ id: String, _ deviceName: String) async throws -> Void
    var delete: @Sendable (_ id: String) async throws -> Void
}

extension PasskeyRepository: TestDependencyKey {
    static let testValue = PasskeyRepository()
    static let previewValue = PasskeyRepository(
        credentials: { [.preview] },
        registrationOptions: {
            PasskeyRegistrationOptions(
                challenge: "",
                rpId: "unes.forcetower.dev",
                userId: "",
                userName: "mariana.nogueira",
                userDisplayName: "Mariana Nogueira",
                excludedCredentialIds: [],
                algorithms: [-7, -257]
            )
        },
        register: { _, _ in },
        rename: { _, _ in },
        delete: { _ in }
    )
}

extension DependencyValues {
    var passkeyRepository: PasskeyRepository {
        get { self[PasskeyRepository.self] }
        set { self[PasskeyRepository.self] = newValue }
    }
}
